README
pam-dbus
========

This simple pam module will, when being used to authenticate a (typically 
remote) user, use dbus to ask any currently logged in (typically local)
user if that is ok.

It connects to the object "de.nomeata.pam_dbus" at the path "/de/nomeata/pam_dbus",
expecting the "de.nomeata.pam_dbus" interface and then call the method "MayLogin".
It passes four strings (service, user, rhost, ruser), some of which might be NULL
and expectes one boolean value, which should be true if the remote user may login.

It has a timeout of 5 minutes after which the UI may stop waiting for a reply.

Configuration
=============

If you want all access to a certain service, (e.g ssh) to be handled by pam_dbus,
replace pam_unix (or @include common-auth on Debian systems) by

>>>>>>>>>snip
auth required pam_dbus.so
<<<<<<<<<<snip

in /etc/pam.d/ssh.

A more advanced configuration can be achieved with pam_succeed_if. The following
snippet, if put instead of "@include common-auth", will make all login requests 
for the user "gast" be handled by libpam_dbus, while every other request will be
handled as usual:

>>>>>>>>>snip
# gast may login using pam-dbus
auth [default=2 success=ignore] pam_succeed_if.so quiet user = gast
auth [success=ok new_authtok_reqd=done ignore=ignore default=die] pam_dbus.so
auth [default=1] pam_permit.so

# Standard Un*x authentication.
@include common-auth
<<<<<<<<<<snip

Note that you need to set "UsePAM yes" in /etc/ssh/sshd_config to make this actually
work.

DBus-Interface
==============

<?xml version="1.0" encoding="UTF-8" ?>

<node name="/de/nomeata/pam_dbus">

  <interface name="de.nomeata.pam_dbus">
    <method name="MayLogin">
      <arg type="s" name="service" direction="in" />
      <arg type="s" name="user" direction="in" />
      <arg type="s" name="rhost" direction="in" />
      <arg type="s" name="ruser" direction="int" />
      <arg type="b" name="login_ok" direction="out" />
    </method>
  </interface>
</node>

Credits
=======

pam_dbus was written by Joachim Breitner <mail@joachim-breitner>, during DebConf8,
having an application on the OpenMoko Freerunner in mind.


debian/changelog
pam-dbus (0.2.1-3) unstable; urgency=medium

  * Build with autoreconf (Closes: #734825) 

 -- Joachim Breitner <nomeata@debian.org>  Tue, 14 Jan 2014 23:34:10 +0000

pam-dbus (0.2.1-2) unstable; urgency=low

  * Build with autotools_dev (Closes: #727486)

 -- Joachim Breitner <nomeata@debian.org>  Thu, 24 Oct 2013 09:55:59 +0200

pam-dbus (0.2.1-1) unstable; urgency=low

  * New upstream release, Closes: #675535

 -- Joachim Breitner <nomeata@debian.org>  Sat, 02 Jun 2012 12:06:39 +0200

pam-dbus (0.2-2) unstable; urgency=low

  * Do not install .la files (Closes: #633198)

 -- Joachim Breitner <nomeata@debian.org>  Sun, 10 Jul 2011 12:35:49 +0200

pam-dbus (0.2-1) unstable; urgency=low

  * New upstream release
    + uses libtool (hopefully closes: #504051)

 -- Joachim Breitner <nomeata@debian.org>  Fri, 31 Oct 2008 00:35:16 +0100

pam-dbus (0.1-1) unstable; urgency=low

  * Initial release

 -- Joachim Breitner <nomeata@debian.org>  Sun, 07 Sep 2008 19:39:01 +0200


bootstrap.sh
#!/bin/sh
# $Id: bootstrap.sh 13 2003-07-31 12:40:54Z lennart $

# This file is part of pam_dbus.
#
# pam_dbus is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# pam_dbus is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with pam_dbus; if not, write to the Free Software Foundation,
# Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.

if [ "x$1" = "xam" ] ; then
    set -ex
    automake -a -c
    ./config.status
else 
    set -ex

    make maintainer-clean || true

    rm -rf autom4te.cache
    rm -f config.cache

    aclocal
    libtoolize -c --force
    autoheader
    automake -a -c
    autoconf -Wall

    ./configure "$@"
fi



handlers/pam-dbus-notification-daemon.py
#!/usr/bin/python

import pygtk
import gtk
import gobject

import dbus
import dbus.service
import dbus.glib

import pynotify

# list of Notifications around, kept to prevent the garbage collector to handle them.
ns = []

def notify_cb(n, action, pam_cb):
    if pam_cb is not None:
        if action == "permit":
            pam_cb(True)
        else:
            pam_cb(False)
        pam_cb = None
    ns.remove(n)

# Initializite pynotify
if not pynotify.init("pam-dbus"):
    sys.exit(1)

# dbus object to answer the request
class PamDbusAnswer(dbus.service.Object):
     def __init__(self, bus_name, object_path="/de/nomeata/pam_dbus"):
         dbus.service.Object.__init__(self, bus_name, object_path)
 
     @dbus.service.method(dbus_interface='de.nomeata.pam_dbus',
                          in_signature='ssss', out_signature='b',
                          async_callbacks=('ret_cb', 'error_cb'))
     def MayLogin(self, service, user, rhost, ruser, ret_cb, error_cb):

         n = pynotify.Notification("Login request","Someone " + 
                            ("named %s " % ruser if ruser else "") +
                            "tries to log in via %s " % service + 
                            ("from %s " % rhost if rhost else "") + 
                            "as user %s. " % user + 
                            "Do you want to allow that?")

         n.set_timeout(5*60*1000)
         n.add_action("permit", "Permit login", notify_cb, ret_cb)
         n.add_action("deny", "Deny login", notify_cb,ret_cb)
         n.connect("closed",lambda n: notify_cb(n,"closed", ret_cb))

         if not n.show():
             print "Failed to send notification"
             sys.exit(1)

         ns.append(n)
 
system_bus = dbus.SystemBus()
name = dbus.service.BusName("de.nomeata.pam_dbus", bus=system_bus)
object = PamDbusAnswer(name)

gtk.main()



src/pam_dbus.c
/***
  This file is part of pam_dbus.

  pam_dbus is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.
  
  pam_dbus is distributed in the hope that it will be useful, but
  WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  General Public License for more details.
  
  You should have received a copy of the GNU General Public License
  along with pam_dbus; if not, write to the Free Software
  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
  USA
***/
        
#define PAM_SM_AUTH

#include <security/pam_modules.h>

#include <glib.h>
#include <dbus/dbus.h>
#include <dbus/dbus-glib.h>

PAM_EXTERN int pam_sm_authenticate(pam_handle_t *ph, int flags, int argc, const char **argv) {
  DBusGConnection *connection;
  GError *error;
  DBusGProxy *proxy;

  gboolean login_ok;
  
  g_type_init ();

  error = NULL;
  connection = dbus_g_bus_get (DBUS_BUS_SYSTEM, &error);
  if (connection == NULL) {
      g_error_free (error);
      return PAM_AUTHINFO_UNAVAIL;
    }

  /* Create a proxy object for the "bus driver" (name "org.freedesktop.DBus") */
  
  proxy = dbus_g_proxy_new_for_name (connection,
                                     "de.nomeata.pam_dbus",
				     "/de/nomeata/pam_dbus",
				     "de.nomeata.pam_dbus"); 

  char *service, *user, *rhost, *ruser;
  pam_get_item(ph, PAM_SERVICE, (const void **)&service);
  pam_get_item(ph, PAM_USER,    (const void **)&user);
  pam_get_item(ph, PAM_RHOST,   (const void **)&rhost);
  pam_get_item(ph, PAM_RUSER,   (const void **)&ruser);

  /* Call ListNames method, wait for reply */
  error = NULL;
  if (!dbus_g_proxy_call_with_timeout (proxy, "MayLogin", 5*60*1000, &error,
			G_TYPE_STRING, service,
			G_TYPE_STRING, user,
			G_TYPE_STRING, rhost,
			G_TYPE_STRING, ruser,
  			G_TYPE_INVALID,
                        G_TYPE_BOOLEAN, &login_ok,
			G_TYPE_INVALID))
    {
      /* Just do demonstrate remote exceptions versus regular GError */
      g_error_free (error);
      g_object_unref (proxy);
      return PAM_AUTHINFO_UNAVAIL;
    }

  /* Print the results */


  g_object_unref (proxy);

  return login_ok ? PAM_SUCCESS : PAM_AUTH_ERR;
}
